File: python/VCFProgenyImputed.py
from __future__ import annotations

# coding: utf-8
# VCFProgenyImputed.py
# 片親がknownで後代がimputeされている家系を補完する
# 2サンプルのみ

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from GenoRecord import GenoRecord
from VCFFamily import *
from ParentImputerByProgeny import *


#################### VCFProgenyImputed ####################

class VCFProgenyImputed(VCFFamilyBase):
	def __init__(self, samples: list[str], records: list[VCFFamilyRecord],
					ref_haps: list[list[int]],
					is_mat_known: bool, map_: Map, vcf: VCFSmall):
		VCFFamilyBase.__init__(self, samples, vcf)
		self.imputer = ParentImputerByProgeny(records, ref_haps,
												is_mat_known, map_, 0.01)
		self.records: list[VCFFamilyRecord] = records
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	def impute(self) -> None:
		self.imputer.impute()

File: src/VCFProgenyImputed.cpp
#include <cmath>
#include "../include/VCFProgenyImputed.h"
#include "../include/common.h"

using namespace std;

VCFProgenyImputed::VCFProgenyImputed(const STRVEC& s,
							const std::vector<VCFFamilyRecord *>& rs,
							const std::vector<std::vector<int>>& ref_haps,
							bool should_impute_mat_, const Map& map_,
							double w, const VCFSmall *vcf) :
				VCFImputable(s, vcf),
				records(rs),
				should_impute_mat(should_impute_mat_),
				imputer(records, ref_haps, should_impute_mat, map_, w) { }

VCFProgenyImputed::~VCFProgenyImputed() {
	Common::delete_all(records);
}

STRVEC VCFProgenyImputed::imputed_samples() const {
	if(should_impute_mat)
		return { samples[0] };
	else
		return { samples[1] };
}

void VCFProgenyImputed::impute() {
	imputer.impute();
}
